Add support for display-line-numbers-mode

rye opened the pull request:
This PR moves `linum-c` to `line-numbers-c` to make it more general-purpose, and extracts existing logic into a helper function before adding another helper method.

The new functionality on 26.1 is to disable `global-linum-mode` (probably not necessary) before enabling `global-display-line-numbers-mode`.  Then, hooks are added to turn off `display-line-numbers-mode` on a few modes that don't seem to like it, such as `dired-mode`, `git-commit-mode`, `magit`-derived modes, and `package-menu-mode`.

Open to further requests or stylistic feedback.  Otherwise, I would say ToT is ready to use in production.

Resolves #98.

rye committed:
Move linum-c to line-numbers-c

This will make this file more general-purpose.

rye committed:
Extract linum logic into its own helper function

This will help to modularize things.

rye committed:
Add support for display-line-numbers on Emacs >= 26.1

Here we add a helper function, kotct/display-line-numbers--turn-off
which can be used in place of lambdas in hooks to turn stuff
off. (From a cursory search I was unable to find an equivalent to
linum-off for display-line-numbers.)

Also, add a helper method which disables global-linum-mode and enables
global-display-line-numbers-mode.  I was considering simply turning
this on in a prog-mode hook, but this seems better.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>
Tested-by: Kristofer Rye <kristofer.rye@gmail.com>

[GitHub] This commit resolves #98.

cg505 commented on .emacs.d/lisp/visual/line-numbers-c.el:
```diff
@@ -0,0 +1,44 @@
+    (setq linum-format (lambda (line)
+                         (let* ((w (length (number-to-string (count-lines (point-min) (point-max)))))
+                                (thing (concat " %" (number-to-string (max 2 w)) "d ")))
+                           (propertize (format thing line) 'face 'linum))))
```
not for this PR, but I think this lambda is causing a lot of performance issues in weird cases

rye replied:
I do too.

cg505 commented on .emacs.d/lisp/visual/line-numbers-c.el:
```diff
@@ -0,0 +1,44 @@
+
+(defun kotct/display-line-numbers--turn-off ()
+  "Turns off `display-line-numbers-mode' for the current buffer."
+  (display-line-numbers-mode -1))
```
... why? `(kotct/display-line-numbers--turn-off)` is literally more typing than `(display-line-numbers-mode -1)`

cg505 replied:
Oh I see, it's for the hooks. Comments there.

rye replied:
Yes, that's exactly why.

cg505 commented on .emacs.d/lisp/visual/line-numbers-c.el:
```diff
@@ -0,0 +1,44 @@
+    (add-hook 'dired-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'git-commit-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'magit-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'package-menu-mode-hook #'kotct/display-line-numbers--turn-off)))
```
A global function seems a little overkill for this simple thing, but idk. I might let-bind a lambda and use that instead, or dolist over the hooks and put the lambda inline (as in
```elisp
(dolist (hook '(dired-mode-hook ...))
  (add-hook hook (lambda () (display-line-numbers-mode -1))))
```
) but this seems fine. Just because it's not how I would do it doesn't make it wrong.

rye replied:
I'm just thinking if we end up wanting to toggle…

My big hangup on adding lambdas over functions is that I feel like this makes it easier to potentially _remove_ hooks, whereas if you set lambdas it's much harder to take care of.  Since the lambdas are doing something, and that something is something that has a pretty descriptive name, I feel like this is the best approach. (I'd obviously not do this if we had special handling such as relative line numbering for `git-commit-mode` or something ridiculous like that.)

rye replied:
Actually, I do have an idea.  Couldn't we just do `(add-hook hook #'kotct/turn-off-line-numbers)` or whatever that is inside that enclosure?  If so, I think that's more elegant.  (More DRY.)

cg505 replied:
>this makes it easier to potentially remove hooks, whereas if you set lambdas it's much harder to take care of

This is a good point actually. I hadn't thought of this.

cg505 reviewed (dismissed):
The thing that really frustrates me about this mode is that there are spaces on both sides of the numbers in addition to the fringe. Just sitting there wasting space for no reason. Unfortunately, this is hard-coded into the Emacs C source (wtf???), so there is not much we can do there. We can, however, fix the `version<` thing. (Forgot to add this to the review, it is a single comment below.) Thanks for taking care of this.

cg505 commented:
Forgot to hit "add" on one of the review comments, sorry. One second.

cg505 commented on .emacs.d/lisp/visual/line-numbers-c.el:
```diff
@@ -0,0 +1,44 @@
+    (add-hook 'magit-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'package-menu-mode-hook #'kotct/display-line-numbers--turn-off)))
+
+(if (not (version< emacs-version "26.1"))
```
So I know I did this on the packup PR, but apparently it is bad style. You should check `(fboundp 'display-line-numbers-mode)` instead.

rye committed:
Use fboundp to determine whether display-line-numbers is available

This makes a lot more sense, because it's actually checking symbols
rather than making assumptions based on a version number.

rye committed:
Use dolist instead of repeated add-hooks

Functionally identical, but much more elegant.

Tested-by: Kristofer Rye <kristofer.rye@gmail.com>

cg505 reviewed (approved):
sick bro

rye commented:
Merging.
